swagger_templates/python/rest.mustache

```diff
@@ -0,0 +1,342 @@
+            )
+
+        post_params = post_params or {}
+        headers = headers or {}
+
+        if 'Content-Type' not in headers:
+            headers['Content-Type'] = 'application/json'
+
+        try:
+            # For `POST`, `PUT`, `PATCH`, `OPTIONS`, `DELETE`
+            if method in ['POST', 'PUT', 'PATCH', 'OPTIONS', 'DELETE']:
+                if headers['Content-Type'] == 'application/json':
+                    r = self._submit_requests(method,
+                                              url,
+                                              query_params=query_params,
+                                              headers=headers,
+                                              body=json.dumps(body))
+                if headers['Content-Type'] == 'application/x-www-form-urlencoded':
+                    r = self._submit_requests(method,
+                                              url,
+                                              query_params=query_params,
+                                              headers=headers,
+                                              post_params=post_params,
+                                              encode_multipart=False)
+                if headers['Content-Type'] == 'multipart/form-data':
+                    # must del headers['Content-Type'], or the correct Content-Type
+                    # which generated by urllib3 will be overwritten.
+                    del headers['Content-Type']
+                    r = self._submit_requests(method,
+                                              url,
+                                              query_params=query_params,
+                                              headers=headers,
+                                              post_params=post_params,
+                                              encode_multipart=True)
+            # For `GET`, `HEAD`
+            else:
+                r = self._submit_requests(method,
+                                          url,
+                                          query_params=query_params,
+                                          headers=headers)
+        except (urllib3.exceptions.SSLError, urllib3.exceptions.HTTPError) as e:
+            msg = "{0}\n{1}".format(type(e).__name__, str(e))
+            raise ApiException(status=0, reason=msg)
+
+        r = RESTResponse(r)
+
+        # In the python 3, the response.data is bytes.
+        # we need to decode it to string.
+        if sys.version_info > (3,):
+            r.data = r.data.decode('utf8')
+
+        # log response body
+        logger.debug("response body: %s" % r.data)
+
+        if r.status not in range(200, 206):
+            raise ApiException(http_resp=r)
+
+        return r
+
+
+    def _submit_requests(self, method, url, query_params=None, headers=None,
+                         body=None, post_params=None, encode_multipart=None):
+        if re.search(r'^http[s]?://localhost') is not None:
```

Comment:
Not a very robust way to check that you are trying to run against the local cluster, for example, calls could come in with 127.0.0.1 as the address instead of the string localhost.  I suppose if you know the caller will always use 'localhost', than this would be ok.

Is it just me or is this not a valid call to re.search?  Shouldn't it be
```
re.search(<pattern>, <string>)
```

Reply:
We actually discussed this in slack prior to submitting this PR. The conclusion of our discussion was by restricting it to localhost then that leaves open the possibility of creating an application that runs on cluster and it forced to provide authentication credentials by using 127.0.0.1 instead of localhost and since we kind of wanted this "run-on-cluster-using-user-session-credentials" feature of the SDK to be only used for internally developed tools, we kind of wanted it to be somewhat difficult to use. Here's a cut and paste of the conversation we had on slack about it, obviously we are open to input and changing on this:

Here's a link to the review

Reply:
The other thing I had proposed in that discussion was to use papi://pipe or perhaps papi://PAPI_SOCK_PATH instead of https://localhost:8080 to indicate to the underlying code that the connection should go through the PAPI_SOCK_PATH, but rather than make that decision in the slack discussion, I figured we could just go with what we have and get your feedback. The advantage to that, as I see it anyway, is that it is very intentional, and thus is easier to interpret and understand when reading the code and I think it further promotes the idea that we only want internally developed tools to make use of this feature because without knowledge of the internal code you would not be able to accidentally start using it like you would with https://localhost:8080.

Reply:
Ok thanks for the backstory.  Yes, I like the idea of using a special URL resource type, I think papi://<uri> is reasonable.  Technically no need to put an extra 'pipe' keyword or the actual socket path in the URL especially if you are just removing it for the call to isi.rest.send_rest_request() but maybe that makes the client code make more sense.

I still don't see how that is a valid re.search() call though, can you comment on that?

Reply:
The issue you found in the re.search() call is a bug.

Reply:
For clarification, I was thinking:
```
host = "papi://PAPI_SOCK_PATH"
api_client = isi_sdk_8_0_1.ApiClient(host)
```
So not the actual path, just a string with that in the name to make it more clear as to why the code is branching into the call to isi.rest.send_rest_request(isi.rest.PAPI_SOCKET_PATH,...)